service/src/test/java/bio/terra/workspace/app/configuration/external/controller/ReferencedGcpResourceControllerTest.java

```diff
@@ -0,0 +1,217 @@
+  public void createBqDatasetReferencedResource() throws Exception {
+    UUID workspaceId = createWorkspace(mockMvc, objectMapper).getId();
+    ApiCreateGcpBigQueryDatasetReferenceRequestBody requestBody =
+        makeGcpBqDatasetReferenceRequestBody();
+    var request = objectMapper.writeValueAsString(requestBody);
+
+    String serializedResponse =
+        createReferencedResourceAndGetSerializedResponse(
+            workspaceId, request, REFERENCED_GCP_BIG_QUERY_DATASET_V1_PATH_FORMAT);
+
+    var retrievedResource =
+        objectMapper.readValue(serializedResponse, ApiGcpBigQueryDatasetResource.class);
+    assertReferenceResourceCommonFields(requestBody.getMetadata(), retrievedResource.getMetadata());
+    assertEquals(
+        requestBody.getDataset().getDatasetId(), retrievedResource.getAttributes().getDatasetId());
+    assertEquals(
+        requestBody.getDataset().getProjectId(), retrievedResource.getAttributes().getProjectId());
+  }
+
+  @Test
+  public void createBqDatatableReferencedResource() throws Exception {
+    UUID workspaceId = createWorkspace(mockMvc, objectMapper).getId();
+    ApiCreateGcpBigQueryDataTableReferenceRequestBody requestBody =
+        makeBqDataTableReferenceRequestBody();
+    var request = objectMapper.writeValueAsString(requestBody);
+
+    String serializedResponse =
+        createReferencedResourceAndGetSerializedResponse(
+            workspaceId, request, REFERENCED_GCP_BIG_QUERY_DATA_TABLE_V1_PATH_FORMAT);
+
+    var retrievedResource =
+        objectMapper.readValue(serializedResponse, ApiGcpBigQueryDataTableResource.class);
+    assertReferenceResourceCommonFields(requestBody.getMetadata(), retrievedResource.getMetadata());
+    assertEquals(
+        requestBody.getDataTable().getDataTableId(),
+        retrievedResource.getAttributes().getDataTableId());
+    assertEquals(
+        requestBody.getDataTable().getDatasetId(),
+        retrievedResource.getAttributes().getDatasetId());
+    assertEquals(
+        requestBody.getDataTable().getProjectId(),
+        retrievedResource.getAttributes().getProjectId());
+  }
+
+  @Test
+  public void createGitRepoReferencedResource() throws Exception {
+    UUID workspaceId = createWorkspace(mockMvc, objectMapper).getId();
+    ApiCreateGitRepoReferenceRequestBody requestBody = makeGitRepoReferenceRequestBody();
+    var request = objectMapper.writeValueAsString(requestBody);
+
+    String serializedResponse =
+        createReferencedResourceAndGetSerializedResponse(
+            workspaceId, request, REFERENCED_GIT_REPO_V1_PATH_FORMAT);
+
+    var retrievedResource = objectMapper.readValue(serializedResponse, ApiGitRepoResource.class);
+    assertReferenceResourceCommonFields(requestBody.getMetadata(), retrievedResource.getMetadata());
+    assertEquals(
+        requestBody.getGitrepo().getGitRepoUrl(),
+        retrievedResource.getAttributes().getGitRepoUrl());
+  }
+
+  private String createReferencedResourceAndGetSerializedResponse(
```

Comment:
It's not useful to return String serializedResponse. Caller will have to convert to ApiGitRepoResource. Do this instead

```
ApiGitRepoResource createReferencedGitRepo() {
}
```

It's better to write `createReferencedGitRepo()` once, than make callers convert to ApiGitRepoResource every time.

Reply:
i want to have one sharable method for all referenced resource creation because the logic is are almost the same. It's up to the caller what to convert to depending on the resource. There's only one creategitrepo test so it seems not that helpful to have a separate `createReferencedGitRepo`.